### PR TITLE
Remove ECast node from the Painless user tree

### PR DESCRIPTION
**Pull request by jdconrad**
This change removes the ECast node from the user tree, so that there are no structural changes to the user tree making it structurally immutable. All generated casts are now built directly into the ir tree instead. This also removes the `isNull` member variable from AExpression because now an instanceof operator can be used instead of having to pass this information between potentially multiple levels of user tree nodes.

Relates to #49869

### CODE REVIEW

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/GeneralCastTests.java [hunk 1]
```diff
@@ -166,9 +166,9 @@ public void testMethodCallDef() {
      * Currently these do not adopt the argument value, we issue a separate cast!
      */
     public void testArgumentsDef() {
-        assertEquals(5, exec("def x = 5L; return (+(int)x);"));
+        //assertEquals(5, exec("def x = 5L; return (+(int)x);"));
```
Comment (rjernst) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/GeneralCastTests.java]: Shouldn't these be uncommented?

**Review by rjernst — APPROVED** [review]
Looks good, a couple minor comments.

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/GeneralCastTests.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/test/java/org/elasticsearch/painless/GeneralCastTests.java]: Oops! Uncommented.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java [hunk 2]
```diff
@@ -83,10 +84,7 @@
      */
     boolean internal = false;
 
-    /**
-     * Set to true by {@link ENull} to represent a null value.
-     */
-    boolean isNull = false;
+    PainlessCast cast = null;
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java]: Added.

**Comment by jdconrad** [comment]
@rjernst Thanks for the review! Will merge as soon as CI passes.

**Pull request merged**